### DISCUSSION TITLE
ABS Native Windows requirements

### DISCUSSION THREAD

**Discussion by mikiher** [advplyr/audiobookshelf#2352] [General]
Hi. My experience installing ABS as a Docker container on Windows was laden with problems:
- Docker Desktop installation issues (had some weird issues with misplaced SSL certificates)
- Docker compose relatively easy, but probably too complex for novices
- Docker context not set up properly for some reason. Manual switching of context was required
- Docker bind volumes very slow, and ABS library watcher does not work with them
- Docker (and/or WSL) default memory footprint very high. Manual tweaking was required
- Mighty Docker crashes/hanging (suddenly unexplainably losing access to a bind volume)
- Update to the latest version not hard, but again, quite complex for novices.

In short, I believe the expertise bar required for reaching a proper setup with Docker on Windows is too high.

I wanted to solicit some advice as to what's currently missing in order to release this great software on Windows without requiring Docker, and I'll try to work on that (if there's no one else doing it). From my perspective, I see the following:
- Build and release scripts (some of them seem to already exist in package.json)
- Installer
  - installing a pre-built executable
  - setting up the config and metadata directories
  - setting up environment variables
  - setting up ffmpeg
  - setting up a service (optional?)
- A tray application (optional?) (similar to Plex)
  - open the app in a new chrome tab
  - check for updates
  - notifies about updates

What else?

**Comment by advplyr** [comment]
Your help on this would be greatly appreciated. I got as far as building the executable for the node server and setting up a tray app.

The tray app I put on github for sharing my effort but doesn't need to be continued, a new project can be setup if there is a better solution. It is here https://github.com/advplyr/audiobookshelf-win

There are technically 2 dependencies, ffmpeg and tone, however we can disregard tone since I plan on removing this. Tone is used for embedding metadata but we can use ffmpeg for that. Ffmpeg has limitations for writing metadata but not enough to justify a separate dependency.

Building the installer is where things got difficult for me. My original plan was to have the installer install ffmpeg and to build the tray app so that it can be updated. This would make it so that you can update the tray application through the GUI as well as update the server version.

I setup the tray app to check for updates of the server in a github repository. For a while I was building the exe and putting it in the release assets https://github.com/advplyr/audiobookshelf/releases/tag/v2.1.4. I stopped because it was confusing a lot of users who thought this was a production ready windows release. Those exe's do work as long as you have ffmpeg installed.

The tray app is starting those exe's in the background and passing in arguments for port number, file paths, etc. https://github.com/advplyr/audiobookshelf-win/blob/master/WinApp/audiobookshelf/TrayOnly.cs#L267

Just to reiterate, we don't need to use that tray app I built if you have a better approach and it may be cleaner to start from scratch. Let me know what you think

**Comment by mikiher** [comment]
I'll take a look at what you started and I'll let you know.
I also need to get up to speed on writing Windows installers - the last one I wrote was on Windows NT, many, many years ago :)
If anyone has any tools recommendation to do that, I'd be happy to hear.

**Comment by mikiher** [comment]
@advplyr , For ffmpeg installation, which build should we install, the essentials or full?
Also, I guess if ffmpeg already exists, then by default we should not install another version. 
Is there some minimum required version of ffmpeg we should check for?

**Comment by mikiher** [comment]
Another couple of questions:
1. I see that there are two projects in audiobookshelf-win, the one you mentioned in c#, and the other in c++. I assume the c++ is just some old experimentation doing the same thing without using dotnet?
2. Did you choose winforms on purpose? Have you played with one of the other UI frameworks, like WPF or Avalonia? I tried Avalonia and it seemed a bit more human-readable with its xaml configurations (although for this app it seems the UI is very simple, so maybe it doesn't really matter that much).
3. Will you consider going back to putting the executable in the release assets in order for the tray app to find it? or do you think it will still be too confusing for users?

Yet another possible approach would be to just install download and install ffmpeg from within the Audiobookshelf server, if it can't find it. Then the application becomes really portable, and you can just download the executable and put it wherever you want (which also seems intuitive to users), and then no need for an installer at all (also possible to open the default browser to http://localhost:port/ when it's up). What do you think about this? I'm aware this approach is a bit more "rough around the edges". On the other hand, it is more in line with the webby nature of Audiobookshelf.

**Reply by advplyr** [reply]
1. Experimentation. I think I pulled in some code to help keep it running in the background if I'm recalling correctly.
2. I just used winforms because I had used it before a few times a long time ago. I'm happy to try out the new options available.
3. I'm not sure. I was thinking that releases could be published in the windows app repository to keep it separate from the server but then the windows app repository wouldn't have its own separate release notes. I guess the executables could also be named in a way to prevent users from downloading it, not sure what that would be.

I haven't looked into ffmpeg yet but off the top of my head I don't think it is a version thing as much as it is that we need certain codecs. If we could bundle it with the windows install and it isn't too large that would be ideal, then we don't have to worry about all that.

**Reply by mikiher** [reply]
We actually have two options for ffmpeg:
1. We can bundle it. The essentials version weighs about 82MB Unpacked (it's static, so it's self-contained). 
Do we also need ffprobe and ffplay? Each of them also weighs about the same.
2. We can download it from the the web the first time we run Audiobookshelf (if not already installed), and extract it into the same directory the executable is installed in.

Regarding codecs, ffmpeg comes in two build flavors, essentials and full. [This page](https://www.gyan.dev/ffmpeg/builds/#libraries) shows which codecs are in each flavor. It seems to me like all we might ever need is in the essentials build, but please take a look.

**Reply by advplyr** [reply]
We need ffprobe, not ffplay. Ffprobe has always came bundled with the builds I've gotten.

I sort of like the idea of bundling it because it reduces the chance of errors. When I was looking at creating a standalone exe of the winforms app it required bundling .NET which was very large, do you know if there is a way around that?

I'm pretty sure the essentials build is all we need. I took a look and don't recognize libraries in the full build that we are using. I think we should use the essentials build and we can always do our own builds if need be in the future, like Jellyfin does. For the Debian package we are using this build https://johnvansickle.com/ffmpeg/git-readme.txt

We almost certainly don't need the full build since we aren't doing any video encoding/decoding.

**Reply by mikiher** [reply]
> I sort of like the idea of bundling it because it reduces the chance of errors. When I was looking at creating a standalone exe of the winforms app it required bundling .NET which was very large, do you know if there is a way around that?

I just realized I haven't answered this - I don't think there's a way around installing .NET (or, at a minimum, the Desktop component of .NET), if we use the C# code for the tray app. I don't think it has to be bundled, though - I think it can be downloaded and installed as part of the ABS installation process, if it's not already installed (many people likely already have a .Net 6.0 runtime installed, since many other apps depend on it).

The only way around this dependency is to use the C++ app. _I think_ I'd like to avoid that. The code is quite inscrutable, unless you're a hardcore Windows programmer :)

**Reply by mikiher** [reply]
I misspoke. 

After looking a bit more, there seems to be a way around this. I assumed that your C# code actually required .Net 6.0, and there's indeed no way around installing that if the user hasn't already installed it. However, it looks like the code doesn't really require .Net 6.0, and can work with .Net **Framework**, that does come bundled with Windows 10/11.

So after some trials and errors and minor changes to the code, I settled on TargetFramework "net462", which is the .Net Framework version that came with Windows 10 v1607, which was released in 2016. I think it's safe to bet that most of our Windows 10 users are using this or a later version of the OS, so I think that at the moment we can safely use this Target and not worry about installing .Net. I'll update if there are new developments.

I removed all .Net 6+ runtimes from my Windows 10, and it looks like the app loads and works without a problem.

**Reply by advplyr** [reply]
It's probably fine to do that considering before we didn't even have a Windows install option. Perhaps we could offer 2 binaries where one is for Windows with .Net Framework. I'm not sure how big that is

**Comment by mikiher** [comment]
@advplyr hi,
I finally have an alpha version of the installer. It's at [mikiher/audiobookshelf-windows](https://github.com/mikiher/audiobookshelf-windows). I'd really appreciate your comments/suggestions/critique. Added you as a collaborator as well.

**Reply by advplyr** [reply]
Hey thanks I've been checking it out and getting it set up. One issue I came across so far is...
```cs
Registry.GetValue(@"HKEY_CURRENT_USER\Software\Audiobookshelf", "DataDir",
    Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), _appName))
 ```
This returns null because the key "HKEY_CURRENT_USER\Software\Audiobookshelf" doesn't exist in the registry for me.

https://learn.microsoft.com/en-us/dotnet/api/microsoft.win32.registry.getvalue?view=net-8.0#definition

We can use `SetValue` if null

**Comment by mikiher** [comment]
Interesting.
I'm assuming you have not used the installer, since it writes that key and
value.
Also, even if it doesn't exist, I'm perplexed at why this returns null,
since the third parameter of GetValue is the default value which should be
returned if that value does not exist.

Thanks for reporting, I'll investigate further...

**Reply by mikiher** [reply]
OK, I fixed this. 
Looks like the key has to exist for the default value to be considered... bloody convoluted Windows APIs :)
I tested this more thoroughly now and it now seems to work in all cases.

**Reply by advplyr** [reply]
Thanks this is working great. I was able to build the installer and I tested an update where it was already installed.
The only thing I noticed in the installer is I had to replace the SetupIconFile since it had your absolute path.
```bash
SetupIconFile="..\Resources\AppIcon.ico"
```

Originally I thought that the tray app would check for new versions of the server and replace the `audiobookshelf.exe`.
How do you think we should handle updates?

**Reply by mikiher** [reply]
> Thanks this is working great. I was able to build the installer and I tested an update where it was already installed.
The only thing I noticed in the installer is I had to replace the SetupIconFile since it had your absolute path.

Ah, thanks for noting. I'll fix that.

> Originally I thought that the tray app would check for new versions of the server and replace the audiobookshelf.exe.
How do you think we should handle updates?

Since I have to keep the installer synced anyway with new versions of the tray app **and** the server, I think I'm going to download and run the new installer version when it's available (with the latest tray app and server) is detected. Do you think we should do this without asking the user? I'd rather ask before installing a new version. 

I'm working on the code now, should be up in a couple of days.

After that, I was planning to write a Github actions workflow that's triggered by audiobookshelf releases, builds the server executable, the tray app, and the installer, and pushes a new audiobookshelf-windows release automatically.

**Reply by advplyr** [reply]
I think we should ask. Are we able to download and install the updated installer after they click a button without launching the actual installer?

**Reply by mikiher** [reply]
OK, I added manual and automatic check for updates (+download & install, if the user accepts).

I also added Github workflow that automatically builds everything up to and including the installer, drafts a new release, and uploads the installer as an asset. For the time being, I still want to have manual control over the final step of publishing the new release (so I can add release notes, and double check that the installer works), but maybe in the future I'll also add automatic publishing.

Can you please take another look? Can you let me know if there's anything else currently missing/wrong? If not, I will change the latest release designation to "beta". Are you comfortable with changing [the documentation](https://www.audiobookshelf.org/docs#windows-install) to point at this?

Also, can you please go over the readme, and see if it's clear enough (especially the first paragraphs up to the Caveats section that are intended for non-developers)?

**Reply by advplyr** [reply]
Thanks I'm checking it out but not much time until the weekend. There is one piece missing on the server side still that I hope to do this weekend which is removing Tone. With the current server users with the windows install won't be able to embed metadata or make m4b audiobooks, but those are the only limitations.

For the signing I think the donations should be used for that. I opened the donations a few weeks ago on my personal account in anticipation for covering some costs that come up. I'm going to start moving the repositories over to the audiobookshelf organization https://github.com/audiobookshelf and setup github sponsorship there. The repos should be easy to move except for advplyr/audiobookshelf because it is linked to the docker images.

**Reply by advplyr** [reply]
I had a chance to check it out and it looks/works great. The readme is clear to me and the developer instructions are clear but we'll start directing users to it for feedback.

Since I didn't get to replace tone this weekend what I'm probably going to do today is hide the audiobook management tools and show a message saying it is temporarily unavailable for the windows install. Once this is done and and server `v2.7.2` is released I will update the website to point users to the beta windows install.

Once we get some feedback and it is stable we should get the installer signed. I've never gone through that process so I don't know what is involved.

**Reply by mikiher** [reply]
Thanks. No rush from my point of view.

**Reply by advplyr** [reply]
I updated the audiobook management tools to show an alert that it is not currently supported on the Windows install.
I published release v2.7.2 but it didn't trigger the github workflow so I just started one manually.

**Reply by advplyr** [reply]
I guess we link to the latest windows installer release in the docs

**Reply by mikiher** [reply]
Yes, please, thanks. I published release 2.7.2 of the windows-installer, and that is what we should point to.

**Reply by mikiher** [reply]
Actually, can you please set the link to https://github.com/mikiher/audiobookshelf-windows/releases/latest so that you don't need to modify it every time there's a new release.

Also, perhaps add a link to the [project page](https://github.com/mikiher/audiobookshelf-windows) itself, so that those precious few who do RTFM can do it :)?

**Reply by advplyr** [reply]
Sounds good I updated the docs

**Comment by mikiher** [comment]
Another question that I was contemplating - the installer is not signed. I looked at the costs of obtaining a code signing certificate, and they look prohibitive (~$300 per year). Maybe you already have such a certificate for Audiobookshelf, and if so, could I use it to sign the installer?

If not, do you have any suggestions on how to proceed with signing (or just forget about it)?

**Comment by dyerseve726** [comment]
I'm not sure if this is the right place to put this, but just wanted to say thank you so much for the native windows version. It's running perfectly for me and showed up right as I was struggling to get it working in Docker. Perfect timing! Great job.

**Reply by mikiher** [reply]
I'm happy it's working well for you.

**Reply by sg501** [reply]
Working great for me as well.

Windows 10 Pro on Xeon e3-1240v5 workstation.

Thanks to everyone for their work on this project.

**Reply by kuldan5853** [reply]
Add +1, haven't seen any issues that are not present on the mainline builds as well. Fantastic work.
